src/evict/evict_lru.c

```diff
@@ -926,28 +930,63 @@ __evict_tune_workers(WT_SESSION_IMPL *session)
 	WT_CACHE *cache;
 	WT_CONNECTION_IMPL *conn;
 	WT_DECL_RET;
-	uint64_t cur_threads, delta_msec, delta_pages, i, target_threads;
+	uint64_t cur_threads, delta_msec, delta_pages, i, target_threads,
+		time_diff;
 	uint64_t pgs_evicted_cur, pgs_evicted_persec_cur;
 	uint32_t thread_surplus;
 
 	conn = S2C(session);
 	cache = conn->cache;
 
 	WT_ASSERT(session, conn->evict_threads.threads[0]->session == session);
-
-	if (conn->evict_tune_stable)
-		return (0);
+	pgs_evicted_cur = pgs_evicted_persec_cur = 0;
 
 	__wt_epoch(session, &current_time);
+	time_diff = WT_TIMEDIFF_SEC(current_time, conn->evict_tune_last_time);
 
 	/*
-	 * Every EVICT_TUNE_PERIOD seconds record the number of
-	 * pages evicted per second observed in the previous period.
+	 * If we have reached the stable state and have not run long enough
+	 * to surpass the forced re-tuning threshold, return.
+	 */
+	if (conn->evict_tune_stable) {
+		if (time_diff < EVICT_FORCE_RETUNE)
+			return (0);
+		else {
+			/* Stable state was reached a long time ago.
+			   Let's re-tune. Reset all the state.*/
+			conn->evict_tune_stable = 0;
+			conn->evict_tune_last_action_time.tv_sec = 0;
+			conn->evict_tune_pgs_last = 0;
+			conn->evict_tune_num_points = 0;
+			conn->evict_tune_pg_sec_max = 0;
+			conn->evict_tune_workers_best = 0;
+
+			/* Reduce the number of eviction workers
+			 * to the minimum */
+			thread_surplus = conn->evict_threads.current_threads -
+				conn->evict_threads_min;
+			for (i = 0; i < thread_surplus; i++) {
+				WT_ERR(__wt_thread_group_stop_one(
+					 session, &conn->evict_threads, false));
+				WT_STAT_CONN_INCR(session,
+				  cache_eviction_worker_removed);
+			}
+			WT_STAT_CONN_INCR(session,
+					  cache_eviction_force_retune);
+		}
+	}
+	/*
+	 * If we have not reached stable state, don't do anything
+	 * unless enough time has passed since the last
+	 * time we have taken any action in this function.
 	 */
-	if (WT_TIMEDIFF_SEC(
-	    current_time, conn->evict_tune_last_time) < EVICT_TUNE_PERIOD)
+	else if (time_diff < EVICT_TUNE_PERIOD)
```

Comment:
the else needs to be in the same line where if block ends. So this block might end up looking like:

```
        if (conn->evict_tune_stable) {
                if (time_diff < EVICT_FORCE_RETUNE)                              
                        return (0);                                                       
                /*                                                               
                 * Stable state was reached a long time ago. Let's re-tune. Reset all the state.
                 */                                                              
                conn->evict_tune_stable = 0;                             
                ....
                ....
        } else if (time_diff < EVICT_TUNE_PERIOD)                                
                /*                                                               
                 * If we have not reached stable state, don't do anything unless 
                 * enough time has passed since the last time we have taken any  
                 * action in this function.                                      
                 */                                                              
                return (0);
```